src/edu/washington/escience/myria/operator/CSVFileScanFragment.java

```diff
@@ -0,0 +1,353 @@
+    this(source, schema, startByteRange, endByteRange, isLastWorker, delimiter, null, null, null);
+  }
+
+  public CSVFileScanFragment(
+      final String filename,
+      final Schema schema,
+      final long startByteRange,
+      final long endByteRange,
+      final boolean isLastWorker,
+      @Nullable final Character delimiter,
+      @Nullable final Character quote,
+      @Nullable final Character escape,
+      @Nullable final Integer numberOfSkippedLines) {
+    this(
+        new FileSource(filename),
+        schema,
+        startByteRange,
+        endByteRange,
+        isLastWorker,
+        delimiter,
+        quote,
+        escape,
+        numberOfSkippedLines);
+  }
+
+  public CSVFileScanFragment(
+      final DataSource source,
+      final Schema schema,
+      final long partitionStartByteRange,
+      final long partitionEndByteRange,
+      final boolean isLastWorker,
+      @Nullable final Character delimiter,
+      @Nullable final Character quote,
+      @Nullable final Character escape,
+      @Nullable final Integer numberOfSkippedLines) {
+    this.source = (AmazonS3Source) Preconditions.checkNotNull(source, "source");
+    this.schema = Preconditions.checkNotNull(schema, "schema");
+
+    this.delimiter = MoreObjects.firstNonNull(delimiter, CSVFormat.DEFAULT.getDelimiter());
+    this.quote = MoreObjects.firstNonNull(quote, CSVFormat.DEFAULT.getQuoteCharacter());
+    this.escape = escape != null ? escape : CSVFormat.DEFAULT.getEscapeCharacter();
+    this.numberOfSkippedLines = MoreObjects.firstNonNull(numberOfSkippedLines, 0);
+
+    this.partitionStartByteRange = partitionStartByteRange;
+    this.partitionEndByteRange = partitionEndByteRange;
+    this.isLastWorker = isLastWorker;
+  }
+
+  @Override
+  protected TupleBatch fetchNextReady() throws IOException, DbException {
+    long lineNumberBegin = lineNumber;
+    boolean fixingStartByte = false;
+    boolean onLastRow = false;
+
+    boolean discardedRecord = false;
+
+    while ((buffer.numTuples() < TupleBatch.BATCH_SIZE)) {
+      discardedRecord = false;
+      lineNumber++;
+      if (parser.isClosed()) {
+        break;
+      }
+      try {
+        if (!iterator.hasNext()) {
+          parser.close();
+          break;
+        }
+      } catch (final RuntimeException e) {
+        throw new DbException("Error parsing row " + lineNumber, e);
+      }
+      CSVRecord record = iterator.next();
+      // This covers the case where the first row of a worker matches the schema. We only want to read this row if the
+      // previous character is '\n' or '\r'
+      if (record.size() == schema.numColumns()
+          && lineNumber - 1 == 0
+          && partitionStartByteRange != 0) {
+        InputStreamReader startStreamReader =
+            new InputStreamReader(
+                source.getInputStream(partitionStartByteRange - 1, partitionEndByteRange));
+        char currentChar = (char) startStreamReader.read();
```

Comment:
It's possible that `read()` will fail if we start in the middle of a multibyte UTF-8 sequence. We should have a test for this (probably easiest just to use something from the top half of Latin-1 charset, i.e., between `U+00F0` and `U+00FF`).